File: Python/CodinGame/Training/classic puzzels - medium/Mars Lander - Episode 2.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;


int main()
{
    int surfaceN; // the number of points used to draw the surface of Mars.
    cin >> surfaceN; cin.ignore();
    int previousY = -1;
    int previousX = -1;
    int landingY, landingX0, landingX1, landingX2;
    for (int i = 0; i < surfaceN; i++) {
        int landX; // X coordinate of a surface point. (0 to 6999)
        int landY; // Y coordinate of a surface point. By linking all the points together in a sequential fashion, you form the surface of Mars.
        cin >> landX >> landY; cin.ignore();
        if (landY == previousY){
            landingY = landY;
            landingX1 = previousX;
            landingX2 = landX;
            landingX0 = (landingX1+landingX2)/2;
            cerr << "LandingX: " << landingX1 << " - " << landingX2 << endl;
        }
        previousX = landX;
        previousY = landY;
    }


    // game loop
    while (1) {
        int X;
        int Y;
        int hSpeed; // the horizontal speed (in m/s), can be negative.
        int vSpeed; // the vertical speed (in m/s), can be negative.
        int fuel; // the quantity of remaining fuel in liters.
        int rotate; // the rotation angle in degrees (-90 to 90).
        int power; // the thrust power (0 to 4).
        cin >> X >> Y >> hSpeed >> vSpeed >> fuel >> rotate >> power; cin.ignore();

        int angel = 0;
        int thrust = 3;
        int diff = X - landingX0;
        int height = Y - landingY;

        cerr << "diff: " << diff << " height: " << height << endl;

        if (X < landingX1 || X > landingX2) //nicht über Landeplatz
        {
            if (vSpeed < 0) thrust = 4;

            //beschleunigen
            if (diff < 0) angel = -30;
            else angel = 30;

            //bremsen
            if (hSpeed > 50 || (diff < -1500 && hSpeed >= 20)) {
                angel = 15;
            }
            else if (hSpeed < -50 || (diff < 1500 && hSpeed <= -20)) {
                angel = -15;
            }
        }
        else //Landen
        {
            if (vSpeed <= -40 || (vSpeed < -10 && height < 1000) || abs(hSpeed) > 30){
                thrust = 4;
            } 

            if (hSpeed <= -10) angel = -45;
            else if (hSpeed >= 10) angel = 45;
            if (height < 250) angel = 0;
        }
        if (vSpeed <= -40) angel = 0;

        // rotate power. rotate is the desired rotation angle. power is the desired thrust power.
        cout << angel << " " << thrust << endl;
    }
}

File: Python/CodinGame/Training/classic puzzels - medium/Mars Lander - Episode 2.py
import sys
import math

surface_n = int(raw_input()) 
previous_y = None
previous_x = None
for i in xrange(surface_n):
    land_x, land_y = [int(j) for j in raw_input().split()]
    print >> sys.stderr, land_x, land_y
    if land_y == previous_y:
        landing_y = land_y
        landing_x = (previous_x, land_x)
    previous_y = land_y
    previous_x = land_x

print >> sys.stderr, landing_y, landing_x
landing_x_optimum = (landing_x[0]+landing_x[1])//2
# game loop
while True:
    # h_speed: the horizontal speed (in m/s), can be negative.
    # v_speed: the vertical speed (in m/s), can be negative.
    # fuel: the quantity of remaining fuel in liters.
    # rotate: the rotation angle in degrees (-90 to 90).
    # power: the thrust power (0 to 4).
    x, y, h_speed, v_speed, fuel, rotate, power = [int(i) for i in raw_input().split()]
    print >> sys.stderr, "h_speed: ", h_speed, " v_speed: ", v_speed
    
    angel = 0
    thrust = 3
    diff = x-landing_x_optimum
    height = y-landing_y
    print >> sys.stderr, "diff: ", diff, " height: ", height
    
    if landing_x[0] < x < landing_x[1]: # Landevorgang
        if v_speed < -40 or (v_speed < -10 and height<1000) or abs(h_speed) > 30:
            thrust = 4
        
        if h_speed < -10:
            angel = -45
        elif h_speed > 10:
            angel = 45

        if height < 250:
            angel = 0

    else:
        if v_speed < 0:
            thrust = 4

        if diff<0:
            angel = -30
        else:
            angel = 30

        if abs(h_speed) >= 50:
            angel = 15*h_speed/abs(h_speed)
        if abs(diff) < 1500 and abs(h_speed) >= 20:
            angel = 15*h_speed/abs(h_speed)
            speed = 3


    if v_speed <= -40:
        angel = 0

    # rotate thrust. rotate is the desired rotation angle. thrust is the desired thrust power.
    print "{0} {1}".format(angel, thrust)
